File: web/src/lib/trainingStats.ts
/**
 * `static/models/training.json` — how the Connect 4 checkpoints were made.
 *
 * Written by `ml/export/training_stats.py` in the same pass that exports the
 * ONNX files, straight out of the checkpoints, so it cannot drift from the
 * weights that shipped. Separate from `manifest.json` on purpose: the manifest
 * is contract-fixed (CONTRACTS §5) and the exhibit depends on its shape, while
 * this is dashboard-only and nothing breaks without it.
 *
 * Types mirror the writer. Everything is optional-tolerant on read: a manifest
 * written by an older export step, or no file at all, has to degrade to "no
 * training stats" rather than throw on a page whose whole point is honesty
 * about missing data.
 */

export interface TrainingArchitecture {
	channels: number | null;
	blocks: number | null;
	parameters: number;
}

export interface TrainingRun {
	gamesTrained: number;
	firstCheckpointAt: string;
	lastCheckpointAt: string;
	/** Seconds between the first and last checkpoint — NOT the run's total
	 *  duration, since the first checkpoint is written after its games are
	 *  already played. Labelled as a span wherever it is shown. */
	spanSeconds: number | null;
}

export interface TrainingCheckpoint {
	id: string;
	gamesTrained: number;
	parameters: number;
	savedAt: string;
	checkpointKb: number;
	onnxKb?: number;
}

export interface TrainingStats {
	architecture: TrainingArchitecture;
	hyperparameters: Record<string, number>;
	run: TrainingRun;
	checkpoints: TrainingCheckpoint[];
	/** Metrics the training pipeline does not produce, and why. Rendered as-is:
	 *  a reader is better served by "no tournament has run" than by silence. */
	missing?: Record<string, string>;
}

export interface TrainingStatsFile {
	version: number;
	connect4: TrainingStats | null;
}

/** Narrows the parsed JSON, returning null for anything unusable. */
export function parseTrainingStats(json: unknown): TrainingStats | null {
	if (typeof json !== 'object' || json === null) return null;
	const file = json as Partial<TrainingStatsFile>;
	const c4 = file.connect4;
	if (typeof c4 !== 'object' || c4 === null) return null;
	if (!Array.isArray(c4.checkpoints) || c4.checkpoints.length === 0) return null;
	if (typeof c4.architecture !== 'object' || c4.architecture === null) return null;
	return c4 as TrainingStats;
}

/** `9m 11s`, `1h 04m`, `47s`. Null when there is nothing to report. */
export function formatDuration(seconds: number | null | undefined): string | null {
	if (seconds === null || seconds === undefined || !Number.isFinite(seconds) || seconds < 0) {
		return null;
	}
	const total = Math.round(seconds);
	const hours = Math.floor(total / 3600);
	const minutes = Math.floor((total % 3600) / 60);
	const secs = total % 60;
	if (hours > 0) return `${hours}h ${String(minutes).padStart(2, '0')}m`;
	if (minutes > 0) return `${minutes}m ${String(secs).padStart(2, '0')}s`;
	return `${secs}s`;
}

/** ISO 8601 to a plain `19 Sep 2026` — no time, since the exact second of a
 *  checkpoint write is noise to a reader. Returns null on anything unparseable. */
export function formatDate(iso: string | null | undefined): string | null {
	if (!iso) return null;
	const d = new Date(iso);
	if (Number.isNaN(d.getTime())) return null;
	return d.toLocaleDateString('en-GB', { day: 'numeric', month: 'short', year: 'numeric' });
}

/** `89,219`. */
export function formatNumber(n: number): string {
	return n.toLocaleString('en-US');
}

/** Hyperparameter keys are camelCase in the JSON; show them as words. */
export function humanizeKey(key: string): string {
	return key
		.replace(/([A-Z])/g, ' $1')
		.replace(/^./, (c) => c.toUpperCase())
		.toLowerCase();
}

File: ml/export/training_stats.py
"""Writes `web/static/models/training.json`: how the checkpoints were made.

Deliberately a separate file from `manifest.json`. The manifest is
contract-fixed (CONTRACTS §5) and answers one question — which opponents
exist — and the browser's exhibit depends on that shape. This answers a
different question, for the dashboard: what the network is, what it was
trained with, and how long it took. Nothing here is required to play a game,
so a missing or stale `training.json` degrades to "no training stats" rather
than breaking the exhibit.

Everything is read back out of the checkpoint files themselves, never
hand-written, so it cannot drift from the weights that shipped.

What is NOT in here, and why:

- **Loss curves.** `train.py` prints per-cycle policy and value loss to stdout
  and saves neither. The existing checkpoints therefore have no recoverable
  history, and inventing one is not an option. Persisting it is a change to
  the training loop, and it would only cover future runs.
- **Elo.** `ml/tournament/` is an `__init__.py` advertising two modules that
  were never written (P3-C), so every rating is `null` and the dashboard says
  "not yet rated" rather than showing a number nobody computed.
"""

from __future__ import annotations

import json
from datetime import datetime
from pathlib import Path
from typing import Any

import torch

TRAINING_STATS_VERSION = 1

#: Config keys worth showing, in the order they should be read. Anything else
#: in the checkpoint's config is training-harness detail (paths, worker counts)
#: that tells a reader nothing about the model.
HYPERPARAMETER_KEYS = (
    "simulations",
    "batch_size",
    "learning_rate",
    "games_per_cycle",
    "train_steps_per_cycle",
    "buffer_capacity",
    "dirichlet_epsilon",
    "dirichlet_alpha",
    "temperature_threshold",
    "seed",
)


def _camel(key: str) -> str:
    head, *rest = key.split("_")
    return head + "".join(word.capitalize() for word in rest)


def read_checkpoint_stats(path: Path) -> dict[str, Any]:
    """Architecture, size and provenance for one checkpoint file."""
    path = Path(path)
    data = torch.load(path, map_location="cpu", weights_only=False)
    state = data["model_state_dict"]
    return {
        "id": path.stem,
        "gamesTrained": int(data["games_trained"]),
        "parameters": int(sum(v.numel() for v in state.values())),
        "savedAt": data["timestamp"],
        "checkpointKb": round(path.stat().st_size / 1024),
        "config": dict(data["config"]),
    }


def _wall_clock_seconds(timestamps: list[str]) -> float | None:
    """Seconds between the first and last checkpoint.

    NOT the run's total duration: the first checkpoint is written once its
    games are already played, so whatever happened before it is not visible
    here. Labelled accordingly wherever it is shown.
    """
    if len(timestamps) < 2:
        return None
    parsed = sorted(datetime.fromisoformat(t) for t in timestamps)
    return (parsed[-1] - parsed[0]).total_seconds()


def build_training_stats(
    checkpoint_paths: list[Path], onnx_dir: Path | None = None
) -> dict[str, Any]:
    """Builds the `training.json` payload from ladder checkpoints."""
    rows = [read_checkpoint_stats(p) for p in checkpoint_paths]
    rows.sort(key=lambda r: r["gamesTrained"])

    if not rows:
        return {"version": TRAINING_STATS_VERSION, "connect4": None}

    config = rows[-1]["config"]
    checkpoints = []
    for row in rows:
        entry = {
            "id": row["id"],
            "gamesTrained": row["gamesTrained"],
            "parameters": row["parameters"],
            "savedAt": row["savedAt"],
            "checkpointKb": row["checkpointKb"],
        }
        if onnx_dir is not None:
            onnx_path = Path(onnx_dir) / f"{row['id']}.onnx"
            if onnx_path.exists():
                entry["onnxKb"] = round(onnx_path.stat().st_size / 1024)
        checkpoints.append(entry)

    return {
        "version": TRAINING_STATS_VERSION,
        "connect4": {
            "architecture": {
                "channels": config.get("channels"),
                "blocks": config.get("blocks"),
                "parameters": rows[-1]["parameters"],
            },
            "hyperparameters": {_camel(k): config[k] for k in HYPERPARAMETER_KEYS if k in config},
            "run": {
                "gamesTrained": rows[-1]["gamesTrained"],
                "firstCheckpointAt": rows[0]["savedAt"],
                "lastCheckpointAt": rows[-1]["savedAt"],
                "spanSeconds": _wall_clock_seconds([r["savedAt"] for r in rows]),
            },
            "checkpoints": checkpoints,
            # Stated, not implied by absence: a reader should be told these are
            # missing on purpose rather than left to wonder.
            "missing": {
                "lossHistory": "train.py prints per-cycle loss and does not persist it",
                "elo": "no tournament has run (ml/tournament is unimplemented)",
            },
        },
    }


def write_training_stats(path: Path, stats: dict[str, Any]) -> dict[str, Any]:
    path = Path(path)
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("w", encoding="utf-8") as f:
        json.dump(stats, f, indent=2)
        f.write("\n")
    return stats
